[BUG] JSONB.parseObject error, `create instance error interface java.util.Set, offset 8`

### 问题描述
*简要描述您碰到的问题。*


### 环境信息
 - JDK： openjdk 11
 - Fastjson2：2.0.40
 - Dubbo: 3.2.5
 

### 重现步骤

#### server
```java
java.io.IOException: org.apache.dubbo.common.serialize.SerializationException: com.alibaba.fastjson2.JSONException: create instance error interface java.util.Set, offset 8
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper.handleToIOException(DefaultSerializationExceptionWrapper.java:358)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper.access$000(DefaultSerializationExceptionWrapper.java:28)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper$ProxyObjectInput.readObject(DefaultSerializationExceptionWrapper.java:168)
	at org.apache.dubbo.rpc.protocol.dubbo.DecodeableRpcInvocation.drawArgs(DecodeableRpcInvocation.java:267)
	at org.apache.dubbo.rpc.protocol.dubbo.DecodeableRpcInvocation.decode(DecodeableRpcInvocation.java:168)
	at org.apache.dubbo.rpc.protocol.dubbo.DecodeableRpcInvocation.decode(DecodeableRpcInvocation.java:102)
	at org.apache.dubbo.rpc.protocol.dubbo.DubboCodec.decodeBody(DubboCodec.java:197)
	at org.apache.dubbo.remoting.exchange.codec.ExchangeCodec.decode(ExchangeCodec.java:138)
	at org.apache.dubbo.remoting.exchange.codec.ExchangeCodec.decode(ExchangeCodec.java:92)
	at org.apache.dubbo.rpc.protocol.dubbo.DubboCountCodec.decode(DubboCountCodec.java:60)
	at org.apache.dubbo.remoting.transport.netty4.NettyCodecAdapter$InternalDecoder.decode(NettyCodecAdapter.java:103)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.dubbo.common.serialize.SerializationException: com.alibaba.fastjson2.JSONException: create instance error interface java.util.Set, offset 8
	... 30 common frames omitted
Caused by: com.alibaba.fastjson2.JSONException: create instance error interface java.util.Set, offset 8
	at com.alibaba.fastjson2.reader.ObjectReaderImplList.readJSONBObject(ObjectReaderImplList.java:468)
	at com.alibaba.fastjson2.JSONB.parseObject(JSONB.java:540)
	at org.apache.dubbo.common.serialize.fastjson2.FastJson2ObjectInput.readObject(FastJson2ObjectInput.java:125)
	at org.apache.dubbo.common.serialize.DefaultSerializationExceptionWrapper$ProxyObjectInput.readObject(DefaultSerializationExceptionWrapper.java:166)
	... 27 common frames omitted
Caused by: java.lang.InstantiationException: java.util.Set
	at java.base/java.lang.Class.newInstance(Class.java:571)
	at com.alibaba.fastjson2.reader.ObjectReaderImplList.readJSONBObject(ObjectReaderImplList.java:466)
	... 30 common frames omitted
Caused by: java.lang.NoSuchMethodException: java.util.Set.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3349)
	at java.base/java.lang.Class.newInstance(Class.java:556)
	... 31 common frames omitted
```

#### client
```java
Caused by: org.apache.dubbo.remoting.RemotingException: Fail to decode request due to: RpcInvocation [methodName=xxxxxMethod, parameterTypes=null]
	at org.apache.dubbo.remoting.exchange.support.DefaultFuture.doReceived(DefaultFuture.java:249)
	at org.apache.dubbo.remoting.exchange.support.DefaultFuture.received(DefaultFuture.java:200)
	at org.apache.dubbo.remoting.exchange.support.DefaultFuture.received(DefaultFuture.java:188)
	at org.apache.dubbo.remoting.exchange.support.header.HeaderExchangeHandler.handleResponse(HeaderExchangeHandler.java:65)
	at org.apache.dubbo.remoting.exchange.support.header.HeaderExchangeHandler.received(HeaderExchangeHandler.java:206)
	at org.apache.dubbo.remoting.transport.DecodeHandler.received(DecodeHandler.java:53)
	at org.apache.dubbo.remoting.transport.dispatcher.ChannelEventRunnable.run(ChannelEventRunnable.java:62)
	at org.apache.dubbo.common.threadpool.ThreadlessExecutor$RunnableWrapper.run(ThreadlessExecutor.java:152)
	at org.apache.dubbo.common.threadpool.ThreadlessExecutor.waitAndDrain(ThreadlessExecutor.java:77)
	at org.apache.dubbo.rpc.AsyncRpcResult.get(AsyncRpcResult.java:205)
```

--- a/core/src/main/java/com/alibaba/fastjson2/reader/ObjectReaderImplList.java
+++ b/core/src/main/java/com/alibaba/fastjson2/reader/ObjectReaderImplList.java
@@ -370,7 +370,12 @@ public final class ObjectReaderImplList
         Function builder = this.builder;
         Class listType = this.instanceType;
         if (objectReader != null) {
-            listType = objectReader.getObjectClass();
+            if (objectReader instanceof ObjectReaderImplList) {
+                listType = ((ObjectReaderImplList) objectReader).instanceType;
+                builder = ((ObjectReaderImplList) objectReader).builder;
+            } else {
+                listType = objectReader.getObjectClass();
+            }
 
             if (listType == CLASS_UNMODIFIABLE_COLLECTION) {
                 listType = ArrayList.class;
@@ -441,7 +446,7 @@ public final class ObjectReaderImplList
             list = entryCnt > 0 ? new ArrayList(entryCnt) : new ArrayList();
         } else if (listType == JSONArray.class) {
             list = entryCnt > 0 ? new JSONArray(entryCnt) : new JSONArray();
-        } else if (listType == Set.class || listType == HashSet.class) {
+        } else if (listType == HashSet.class) {
             list = new HashSet();
         } else if (listType == LinkedHashSet.class) {
             list = new LinkedHashSet();

--- a/core/src/test/java/com/alibaba/fastjson2/issues_1800/Issue1857.java
+++ b/core/src/test/java/com/alibaba/fastjson2/issues_1800/Issue1857.java
@@ -1,0 +1,50 @@
+package com.alibaba.fastjson2.issues_1800;
+
+import com.alibaba.fastjson2.JSONB;
+import com.alibaba.fastjson2.JSONReader;
+import com.alibaba.fastjson2.JSONWriter;
+import com.alibaba.fastjson2.filter.ContextAutoTypeBeforeHandler;
+import org.junit.Test;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+public class Issue1857 {
+
+    @Test
+    public void test() {
+        Map<String, String> map = new HashMap<>();
+
+        map.put("key1", "value1");
+        map.put("key2", "value2");
+        map.put("key3", "value3");
+
+        byte[] bytes = JSONB.toBytes(map.keySet(),
+                JSONWriter.Feature.WriteClassName,
+                JSONWriter.Feature.FieldBased,
+                JSONWriter.Feature.ReferenceDetection,
+                JSONWriter.Feature.WriteNulls,
+                JSONWriter.Feature.NotWriteDefaultValue,
+                JSONWriter.Feature.NotWriteHashMapArrayListClassName,
+                JSONWriter.Feature.WriteNameAsSymbol);
+
+        Collection<String> result = JSONB.parseObject(bytes, Collection.class, new Handler(),
+                JSONReader.Feature.UseDefaultConstructorAsPossible,
+                JSONReader.Feature.UseNativeObject,
+                JSONReader.Feature.IgnoreAutoTypeNotMatch,
+                JSONReader.Feature.FieldBased);
+        assertEquals(result, map.keySet());
+    }
+
+    private static class Handler extends ContextAutoTypeBeforeHandler {
+
+        public Handler() {
+            super(true, new String[0]);
+        }
+
+    }
+
+}

--- a/core/src/test/java/com/alibaba/fastjson2/issues_1800/Issue1857.java
+++ b/core/src/test/java/com/alibaba/fastjson2/issues_1800/Issue1857.java
@@ -13,7 +13,6 @@ import java.util.Map;
 import static org.junit.Assert.assertEquals;
 
 public class Issue1857 {
-
     @Test
     public void test() {
         Map<String, String> map = new HashMap<>();
@@ -39,12 +38,10 @@ public class Issue1857 {
         assertEquals(result, map.keySet());
     }
 
-    private static class Handler extends ContextAutoTypeBeforeHandler {
-
+    private static class Handler
+            extends ContextAutoTypeBeforeHandler {
         public Handler() {
             super(true, new String[0]);
         }
-
     }
-
 }

